quantum/config_common.h

```diff
@@ -14,66 +14,116 @@
+#pragma once
 
 /* diode directions */
 #define COL2ROW       0
 #define ROW2COL       1
 #define CUSTOM_MATRIX 2 /* Disables built-in matrix scanning code */
 
 #ifdef __AVR__
+    #ifndef __ASSEMBLER__
+      #include <avr/io.h>
+    #endif
+    #define PORT_SHIFTER 4 // this may be 4 for all AVR chips
+
+    // If you want to add more to this list, reference the PINx definitions in these header
+    // files: https://github.com/vancegroup-mirrors/avr-libc/tree/master/avr-libc/include/avr
+
+    #if defined(__AVR_ATmega32U4__) || defined(__AVR_ATmega16U4__)
+        #define ADDRESS_BASE 0x00
+        #define PINB_ADDRESS 0x3
+        #define PINC_ADDRESS 0x6
+        #define PIND_ADDRESS 0x9
+        #define PINE_ADDRESS 0xC
+        #define PINF_ADDRESS 0xF
+    #elif defined(__AVR_ATmega32U2__) || defined(__AVR_ATmega16U2__)
+        #define ADDRESS_BASE 0x00
+        #define PINB_ADDRESS 0x3
+        #define PINC_ADDRESS 0x6
+        #define PIND_ADDRESS 0x9
+    #elif defined(__AVR_AT90USB1286__) || defined(__AVR_AT90USB646__)
+        #define ADDRESS_BASE 0x00
+        #define PINA_ADDRESS 0x0
+        #define PINB_ADDRESS 0x3
+        #define PINC_ADDRESS 0x6
+        #define PIND_ADDRESS 0x9
+        #define PINE_ADDRESS 0xC
+        #define PINF_ADDRESS 0xF
+    #elif defined(__AVR_ATmega32A__)
+        #define ADDRESS_BASE 0x10
+        #define PIND_ADDRESS 0x0
+        #define PINC_ADDRESS 0x3
+        #define PINB_ADDRESS 0x6
+        #define PINA_ADDRESS 0x9
+    #else
+        #error "Pins are not defined"
+    #endif
+
     /* I/O pins */
-    #ifndef F0
-        #define B0 0x30
-        #define B1 0x31
-        #define B2 0x32
-        #define B3 0x33
-        #define B4 0x34
-        #define B5 0x35
-        #define B6 0x36
-        #define B7 0x37
-        #define C0 0x60
-        #define C1 0x61
-        #define C2 0x62
-        #define C3 0x63
-        #define C4 0x64
-        #define C5 0x65
-        #define C6 0x66
-        #define C7 0x67
-        #define D0 0x90
-        #define D1 0x91
-        #define D2 0x92
-        #define D3 0x93
-        #define D4 0x94
-        #define D5 0x95
-        #define D6 0x96
-        #define D7 0x97
-        #define E0 0xC0
-        #define E1 0xC1
-        #define E2 0xC2
-        #define E3 0xC3
-        #define E4 0xC4
-        #define E5 0xC5
-        #define E6 0xC6
-        #define E7 0xC7
-        #define F0 0xF0
-        #define F1 0xF1
-        #define F2 0xF2
-        #define F3 0xF3
-        #define F4 0xF4
-        #define F5 0xF5
-        #define F6 0xF6
-        #define F7 0xF7
-        #define A0 0x00
-        #define A1 0x01
-        #define A2 0x02
-        #define A3 0x03
-        #define A4 0x04
-        #define A5 0x05
-        #define A6 0x06
-        #define A7 0x07
+    #define PINDEF(port, pin) ((PIN##port##_ADDRESS << PORT_SHIFTER) | pin)
+
+    #ifdef PORTA
+        #define A0 PINDEF(A, 0)
+        #define A1 PINDEF(A, 1)
+        #define A2 PINDEF(A, 1)
```

Comment:
This line is copy-pasted from the previous one but should apply 2 as the pin instead of 1